tmail-backend/integration-tests/jmap/jmap-integration-tests-common/src/main/scala/com/linagora/tmail/james/common/EmailRecoveryActionIntegrationTest.scala

```diff
@@ -0,0 +1,327 @@
+
+    bobBaseRequest = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(BOB, BOB_PASSWORD)))
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build()
+
+    andreBaseRequest = baseRequestSpecBuilder(server)
+      .setAuth(authScheme(UserCredential(ANDRE, ANDRE_PASSWORD)))
+      .addHeader(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+      .build()
+  }
+
+  @Test
+  def recoveryShouldRestoreJmapDeletedEmail(): Unit = {
+    // bob send an email to andre
+    bobSendAnEmailToAndre()
+
+    val andreMessageId: String = awaitAndreAlreadyHasExpectedEmail(expectEmailNumber = 1).head // andre receive the email
+
+    // andre delete the email
+    `given`(andreBaseRequest)
+      .body(
+        s"""{ "using": ["urn:ietf:params:jmap:core", "urn:ietf:params:jmap:mail"],
+           |  "methodCalls": [
+           |      ["Email/set",{
+           |        "accountId": "$ANDRE_ACCOUNT_ID",
+           |        "destroy": ["$andreMessageId"]
+           |      }, "c1"],
+           |      [ "Email/get", {
+           |        "accountId": "$ANDRE_ACCOUNT_ID",
+           |        "ids": ["$andreMessageId"]
+           |      }, "c2" ]] }""".stripMargin)
+    .when()
+      .post()
+    .`then`
+      .statusCode(HttpStatus.SC_OK)
+      .body("methodResponses[0][1].destroyed", Matchers.containsInAnyOrder(andreMessageId))
+      .body("methodResponses[1][1].list", Matchers.empty())
+      .body("methodResponses[1][1].notFound", Matchers.containsInAnyOrder(andreMessageId))
+
+    // andre restore the email by JMAP EmailRecoveryAction/set
+    andrePostEmailRecoveryCreateAction()
+
+    // verify the restore email success (the email is in andre's inbox)
+    await.untilAsserted(() => {
+      `given`(andreBaseRequest)
+        .body(jmapEmailQueryAllOfAndre)
+      .when()
+        .post()
+      .`then`
+        .body("methodResponses[1][1].list[0].subject", Matchers.is("World domination"))
+    })
+  }
+
+  @Test
+  def recoveryShouldNotImpactOtherUsers(): Unit = {
+    bobSendAnEmailToAndre()
+    awaitAndreAlreadyHasExpectedEmail(expectEmailNumber = 1)
+    deleteAllEmail(andreBaseRequest, ANDRE_ACCOUNT_ID)
+    deleteAllEmail(bobBaseRequest, BOB_ACCOUNT_ID)
+    andrePostEmailRecoveryCreateAction()
+    awaitAndreAlreadyHasExpectedEmail(expectEmailNumber = 1)
+
+    // bob should not impact by andre's recovery action
+    await.untilAsserted(() => {
+      `given`(bobBaseRequest)
+        .body(
+          s"""{ "using": ["urn:ietf:params:jmap:core", "urn:ietf:params:jmap:mail"],
+             |  "methodCalls": [[
+             |      "Email/query",
+             |      {
+             |        "accountId": "$BOB_ACCOUNT_ID",
+             |        "filter": {}
+             |      }, "c1"]] }""".stripMargin)
+      .when()
+        .post()
+      .`then`
+        .body("methodResponses[0][1].ids", Matchers.empty())
+    })
+  }
+
+  @Test
+  def recoveryShouldNotBeRemovedFromTheVault(): Unit = {
```

Comment:
Open question: should an email really not be removed after we recovered it?

Reply:
I think we had the debate once in a loong loong time, can't recall the answer.

Reply:
Your mean is we should remove the vault after recovery? or this test is not ensured?
It is based on the `org.apache.james.webadmin.integration.vault.DeletedMessageVaultIntegrationTest#restoredMessagesShouldNotBeRemovedFromTheVault`

Reply:
If it was the original design then (thus the loooong looooong time ago^^) I'm fine with this then. Thanks for the precision :)